generator/internal/discovery_type_vertex.cc

```diff
@@ -13,12 +13,28 @@
 // limitations under the License.
 
 #include "generator/internal/discovery_type_vertex.h"
+#include "generator/internal/codegen_utils.h"
 #include "google/cloud/internal/absl_str_join_quiet.h"
+#include "google/cloud/internal/make_status.h"
 #include "absl/strings/str_format.h"
+#include "absl/types/optional.h"
 
 namespace google {
 namespace cloud {
 namespace generator_internal {
+namespace {
+
+absl::optional<std::string> CheckForScalarType(nlohmann::json const& j) {
+  std::string type;
+  if (j.contains("type")) type = j["type"];
+  if (type == "string") return "string";
+  if (type == "boolean") return "bool";
+  if (type == "integer") return j.value("format", "int32");
+  if (type == "number") return j.value("format", "float");
+  return {};
```

Comment:
Nit: I can never remember if we have a preference between `{}` and `absl::nullopt` for an empty `absl::optional`.  I believe I would say that the latter is clearer.